javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,111 @@ private[lang] object StackTrace {
+  if (writableStackTrace)
+    fillInStackTrace()
 
-  def initCause(cause: Throwable): Throwable = {
-    e = cause
-    this
-  }
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-suppression not permitted")
+    }
+
+    if (enableSuppression) this.synchronized {
+      if (suppressed eq null) {
+        suppressed = Array(exception)
+      } else {
+        val length        = suppressed.length
+        val newSuppressed = new Array[Throwable](length + 1)
+        System.arraycopy(suppressed, 0, newSuppressed, 0, length)
+        newSuppressed(length) = exception
+        suppressed = newSuppressed
+      }
+    }
+  }
 
   def fillInStackTrace(): Throwable = {
-    this.stackTrace = StackTrace.currentStackTrace()
+    // currentStackTrace should be handling exclusion in its own
+    // critical section, but does not. So do
+    if (writableStackTrace) this.synchronized {
+      this.stackTrace = StackTrace.currentStackTrace()
+    }
     this
   }
 
+  def getCause(): Throwable = e
+
+  def getLocalizedMessage(): String = getMessage()
+
+  def getMessage(): String = s
+
   def getStackTrace(): Array[StackTraceElement] = {
     if (stackTrace eq null) {
-      stackTrace = Array.empty
-    }
-    stackTrace
+      Array.empty[StackTraceElement]
+    } else
+      this.synchronized {
+        stackTrace.clone
+      }
   }
 
-  def setStackTrace(stackTrace: Array[StackTraceElement]): Unit = {
-    var i = 0
-    while (i < stackTrace.length) {
-      if (stackTrace(i) eq null)
-        throw new NullPointerException()
-      i += 1
+  final def getSuppressed(): Array[Throwable] = {
+    if (suppressed == null) {
+      Array.empty[Throwable]
+    } else
+      this.synchronized {
+        suppressed.clone()
+      }
+  }
+
+  def initCause(cause: Throwable): Throwable = {
+    // Java 8 spec says initCause has "at-most-once" semantics,
+    // where implied use in a constructor counts.
+
+    if (cause == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-causation not permitted")
     }
 
-    this.stackTrace = stackTrace.clone()
+    this.synchronized {
+      if (e != null) {
+        // Yes, JVM uses contraction "Can't"
+        val msg = if (cause == null) "a null" else cause.toString
+        throw new java.lang.IllegalStateException(
+          s"Can't overwrite cause with ${msg}")
+      } else {
```

Comment:
I was thinking of checking the exception throwing paths and the messages.

Reply:
Two tests in ThrowablesSuite.scala, below, test the paths which throw exceptions and that
the cause Throwable has object equality with the cause Throwable given as an argument
to initCause().  If there is no object equality, the messages within the result & expected cause
throwables are given by the assert().
```
test("initCause(cause) - cases which throw an Exception")
test("initCause(cause)")
```   

Does this resolve your concern?

The test suite code tests that the type of exception thrown is as expected. It does not check that the
text of the message is as expected.  There is code in other suites which does an exact message match,
but it is tedious to implement.

Reply:
That seems totally fine and would port back to Scala.js as well.